### PR TITLE
DM-27141: General improvements and documentation of ts_cycle_build

### PR DESCRIPTION
**Pull request by tribeiro**
Starting to write documentation for the package in project format.
Update Jenkinsfile.cycle to allow building components using Jenkins server.
Update cycle build to 0015.

### CODE REVIEW

File: doc/user-guide/user-guide.rst [hunk 1]
```diff
@@ -1,74 +1,147 @@
...
+  - ts_idl: Contains ancillary files used by ts_salobj and other higher level applications.
+  - ts_salobj: High-level Python library to develop CSCs and SAL Script.
+
+
+.. _Building-a-new-Cycle-or-Revision:
+
+Building a new Cycle or Revision
+================================
+
+There are two ways users can run a build; manually on a private server or using Jenkins server.
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:91]: replace "Jenkins server" with "the TSSW Jenkins server". Add a link to the Jenkins page too.

File: doc/user-guide/user-guide.rst [hunk 2]
```diff
@@ -1,74 +1,147 @@
...
+Since the build requires credentials for the nexus servers, both to access the licensed version of OpenSplice and to pull/push images to the nexus docker registry, it is advisable to use the Jenkins server.
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:92]: Add TSSW in front of Jenkins in this sentence.

File: doc/user-guide/user-guide.rst [hunk 3]
```diff
@@ -1,74 +1,147 @@
...
+
+.. note::
+
+  At this point, cycle and revision are still managed manually by users.
+  We are investigating ways of automating the process so users that have to worry about setting them up.
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:97]: replace "so users that" with "so users don't"

File: doc/user-guide/user-guide.rst [hunk 4]
```diff
@@ -1,74 +1,147 @@
...
+The process to create a new cycle/revision build is to clone the repository and create a ticket branch to work on.
+
+.. prompt:: bash
+
+  git clone https://github.com/lsst-ts/ts_cycle_build.git
+  cd ts_cycle_build/
+  git checkout tickets/DM-XXXXX
+
+Then, update the cycle/revision and software versions as needed.
+If any of the :ref:`core packages <Core-packages>` version is updated, the cycle number must be incremented, otherwise just increment the revision number and updade the software versions.
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:108]: change "updade" to "update"

File: doc/user-guide/user-guide.rst [hunk 5]
```diff
@@ -1,74 +1,147 @@
...
 
-[This area should link to the XML, then discuss the primary commands/events/telemetry that are expected to be used.
-A full explanation of all possible commands/events/telemetry is beyond the scope of this document.
-However, the description in the XML should be verified to be adequately populated such that a user can find the deeper information if required.]
+   Jenkins build with parameters page.
 
-Example Use-Case
-================
+The build is divided into different steps.
+These steps are designed to maximized reusability of docker layer, minimizing the number of layers in the image and reducing the time it takes to build the system.
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:135]: replace "docker layer" with "docker layers"

File: doc/version-history.rst [hunk 6]
```diff
@@ -10,8 +10,45 @@ Version History
...
-------
-* Added Jenkinsfile for conda recipe
-* Added conda recipe
-* Incorporated new offset parameters with the option of them being persistent (sticky)
\ No newline at end of file
+
+Cycle 0015
+==========
+
+Release of salobj 6/sal 5 with xml 6.2.0.
```
Comment (mareuter) [inline doc/version-history.rst:17]: Bump to 6.2.1.

File: doc/user-guide/user-guide.rst [hunk 7]
```diff
@@ -1,74 +1,147 @@
...
+The steps in the build are as follows:
```
Comment (mareuter) [inline doc/user-guide/user-guide.rst:136]: This doesn't look like all of the components are listed here. Do we want to put the list here? Is it better to just point at the Jenkinsfile for this?

**Review by mareuter — CHANGES_REQUESTED** [review]
- A few questions and some wordsmithing in the text
- Using the revision is discussed in the docs but not implemented in the code. Is this correct?
- We should clearly delineate what are core packages and what is not. I'm not clear on whether or not the stack version necessitates a cycle build.
- We should put the packages the necessitate a new cycle build near the head of the cycle.env file and clearly comment this fact.

File: doc/user-guide/user-guide.rst [hunk 8]
```diff
@@ -1,74 +1,147 @@
...
+Even though components are designed as independent actors they are not isolated entities, and must communicate with each other to accomplish higher level of operation.
+For example, the Main Telescope Mount component (``MTMount``) is must be able to track targets in response to data generated by the Main Telescope Pointing component (``MTPtg``).
+In order for this to work properly ``MTPtg`` must have the correct version of the interface of ``MTMount`` or they would not be able to co-operate.
 
-  and
+This issue is intensified by the architecture of communication middleware, Data Distribution Service (DDS).
+DDS relies on strongly typed Topics, e.g. data structures similar to C struct, to exchange data between hosts.
+For hosts to be able to communicate they must know the Topic type beforehand.
+If the version of the Topics on two host is different, whoever tries to subscribe last (be it to read or write) will fail.
+In general this problem can be avoided by use of loosely typed topics and schema evolution, which allows readers and writers to exchange any type of data structure and translate different versions.
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply doc/user-guide/user-guide.rst:34]: I think it is possible... we could have a single "payload" attribute which is a json string that is parsed around..  I think this is sort of what we do with kafka and Avro on the EFD side. kafka itself does not have schema evolution or anything like that, we use Avro for that part.

File: Jenkinsfile.cycle [hunk 9]
```diff
@@ -60,5 +69,53 @@ pipeline {
...
+      when {
+          expression { params.base_components }
+      }
+      steps {
+
+        script{
+          withEnv(["HOME=${env.WORKSPACE}"]) {
+            docker.withRegistry("https://ts-dockerhub.lsst.org/", "nexus3-lsst_jenkins") {
+              sh """
+              docker-compose --env-file cycle/cycle.env -f cycle/docker-compose.yaml build ataos atdome atdometrajectory athexapod atmcs_sim atpneumatics_sim atspectrograph hexapod mtdome mtdometrajectory mtm2_sim mtmount ospl-daemon rotator salkafka watcher
```
[the remark this answers is not in the document]
Comment (tribeiro) [reply Jenkinsfile.cycle:106]: This is a great idea. Do you mind if I implement it in DM-26305 instead? Don't want to create merge conflicts if I can avoid them... :)

**Comment by tribeiro** [comment]
I implemented most of your suggestions. Responding to your review questions:

- Using the revision is discussed in the docs but not implemented in the code. Is this correct?
  - Yes. I am still looking at a good way to implement it. 

- We should clearly delineate what are core packages and what is not. I'm not clear on whether or not the stack version necessitates a cycle build.
  - At this time I was assuming only our core packages would could for the cycle count. So the dm stack could be a revision. But I can see arguments for both. So, maybe something we will have to figure out in the field?

- We should put the packages the necessitate a new cycle build near the head of the cycle.env file and clearly comment this fact.
  - Yes, that is what I current have. I will make a comment on cycle.env about that. But I'd rather do it in DM-27305 if you don't mind. Just to avoid merge conflicts.

File: doc/user-guide/user-guide.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (mareuter) [reply doc/user-guide/user-guide.rst:34]: Interesting idea. There's quite a loss of clarity in what payload means without schema validation being used.

File: doc/user-guide/user-guide.rst (hunk 7 above)
Comment (mareuter) [reply doc/user-guide/user-guide.rst:138]: OK. I think that looks like everyone.

**Comment by mareuter** [comment]
>     * We should put the packages the necessitate a new cycle build near the head of the cycle.env file and clearly comment this fact.
>       
>       * Yes, that is what I current have. I will make a comment on cycle.env about that. But I'd rather do it in DM-27305 if you don't mind. Just to avoid merge conflicts.

#

File: doc/user-guide/user-guide.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (tribeiro) [reply doc/user-guide/user-guide.rst:34]: yes... it would make the topics completely opaque.... not suggesting we do it, just saying it is possible...

**Review by mareuter — APPROVED** [review]
Thanks for the updates. We can wait on the cycle.env work until a more convenient time. PR approved.

**Pull request merged**